interpreter/text/parser.mly

```diff
@@ -842,6 +845,61 @@ memory_fields :
       [{dinit = $3; dmode = Active {index = x; offset} @@ at} @@ at],
       [], [] }
 
+event :
+  | LPAR EVENT bind_var_opt event_fields RPAR
+    { let at = at () in
+      fun c -> let x = $3 c anon_event bind_event @@ at in fun () -> $4 c x at }
+
+event_fields :
+  | type_use event_types
+    { fun c x at ->
+      let y = inline_type_explicit c ($1 c type_) $2 at in
+      [y], [], [] }
+  | event_types  /* Sugar */
+    { fun c x at ->
+      let y = inline_type c $1 at in
+      [y], [], [] }
+  | inline_import type_use event_fields_import  /* Sugar */
+    { fun c x at ->
+      let y = inline_type_explicit c ($2 c type_) $3 at in
+      [],
+      [{ module_name = fst $1; item_name = snd $1;
+         idesc = EventImport y @@ at } @@ at ], [] }
+  | inline_import event_fields_import  /* Sugar */
+    { fun c x at ->
+      let y = inline_type c $2 at in
+      [],
+      [{ module_name = fst $1; item_name = snd $1;
+         idesc = EventImport y @@ at } @@ at ], [] }
+  | inline_export event_fields  /* Sugar */
+    { fun c x at ->
+      let fns, ims, exs = $2 c x at in fns, ims, $1 (EventExport x) c :: exs }
+
+event_fields_import :  /* Sugar */
+  | event_fields_import_result { $1 }
+  | LPAR PARAM value_type_list RPAR event_fields_import
+    { let FuncType (ins, out) = $5 in FuncType ($3 @ ins, out) }
+  | LPAR PARAM bind_var value_type RPAR event_fields_import  /* Sugar */
+    { let FuncType (ins, out) = $6 in FuncType ($4 :: ins, out) }
+
+event_fields_import_result :  /* Sugar */
+  | /* empty */ { FuncType ([], []) }
+  | LPAR RESULT value_type_list RPAR event_fields_import_result
+    { let FuncType (ins, out) = $5 in FuncType (ins, $3 @ out) }
+
+event_types :
```

Comment:
Nit: `event_type`, since it represents a single event type. Also, I think you can define this as simply
```
event_type :
  | func_type
    { fun c -> EventType ($1 c) }
```
(assuming you added the `EventType` constructor as suggested above).

Reply:
Good point, I simplified this even further by removing the new rule completely. It returns the type index, not the actual type, so it can be factorized.